### ISSUE TITLE
Allow to override warning level per item in no-restricted-syntax

### ISSUE THREAD

**Issue by JuroOravec** [eslint/eslint#13831]
**What rule do you want to change?**

`no-restricted-syntax`

**Does this change cause the rule to produce more or fewer warnings?**

Neither, but it will provide more flexibility.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option per syntax item to enable overriding the severity (off / warn / error).

**Please provide some example code that this change will affect:**

Instead of having a single warning level for all syntax items specified:

<details>

<summary>Current no-restricted-syntax</summary>

```js
    'no-restricted-syntax': [
      'error',
      {
        // Raise on `Vue.extend()`
        selector:
          'CallExpression > MemberExpression[object.name="Vue"][property.name="extend"]',
        message: "Use 'defineComponent' instead of 'Vue.extend'.",
      },
      {
        // Raise when calling watch() inside a watch() callback
        selector:
          'CallExpression[callee.name="watch"] CallExpression[callee.name="watch"]',
        message: 'Do not use nested watchers.',
      },
    ],
```

</details>

The suggestion is to allow to specify warning level *per syntax item*. The warning level could be set e.g. by `severity` property.

The `severity` property would be optional. If not set (`undefined`), the warning level for that syntax item would fall back to the level of the whole rule, as set by the first argument.

The value of `severity` property would follow the same logic as when setting the warning level for the whole rule.

In other words, the syntax item would be extended by following:

`{ severity?: 0 | 1 | 2 | 'off' | 'warn' | 'error' }`

This change would be backward-compatible:
- If severity not set, everything is as before
- If the syntax item is set only as a string, everything is as before.

<details>

<summary>Suggested change to no-restricted-syntax</summary>

```js
    'no-restricted-syntax': [
      'error',
      // This rule does not specify the severity, so it defaults to what is specified above - error
      {
        // Raise on `Vue.extend()`
        selector:
          'CallExpression > MemberExpression[object.name="Vue"][property.name="extend"]',
        message: "Use 'defineComponent' instead of 'Vue.extend'.",
      },
      // This rule specifies the severity to error - error
      {
        // Raise when calling watch() inside a watch() callback
        selector:
          'CallExpression[callee.name="watch"] CallExpression[callee.name="watch"]',
        message: 'Do not use nested watchers.',
  		severity: "error"
      },
      // This rule specifies the severity to warn - warn
      {
          "selector": "CallExpression[callee.name='setTimeout'][arguments.length!=2]",
          "message": "setTimeout must always be invoked with two arguments.",
  		  severity: "warn"
      }
    ],
```

</details>

**What does the rule currently do for this code?**

The rule is great way to quickly write custom syntax rules without needing to write a full-blown rule. However, currently all syntaxes raise the same response (off / warn / error).

This limits the usability of this rule. We wanted to try out using the rule for deprecating unwanted syntax in our codebase. For that use case, each restricted syntax item should have own warning level, so they can be deprecated gradually and independently.

**What will the rule do after it's changed?**

The rule will allow to override the warning levels per syntax item.

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by mdjermanovic** [comment]
Hi @JuroOravec, thanks for the issue!

It isn't likely we can accept this enhancement since it would require significant changes in ESLint's core design.

You can make a clone of `no-restricted-syntax` as a custom/plugin rule (tools like [eslint-rule-extender](https://github.com/kaicataldo/eslint-rule-extender) or [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) can help), and then set one rule to `'warn'` and the other to `'error'`.

**Comment by JuroOravec** [comment]
Hey @mdjermanovic thanks for clarification. Indeed cloning existing, or creating new rule might be the best approach here (I found [this article by Kenneth Truyers](https://www.kenneth-truyers.net/2016/05/27/writing-custom-eslint-rules/#:~:text=In%20order%20to%20write%20a,must%20start%20with%20eslint%2Dplugin) to be really useful).

Closing this issue.
